src/replication.c

```diff
@@ -2143,8 +2143,19 @@ void readSyncBulkPayload(connection *conn) {
             if (old_rdb_fd != -1) close(old_rdb_fd);
             return;
         }
+
         /* Close old rdb asynchronously. */
         if (old_rdb_fd != -1) bioCreateCloseJob(old_rdb_fd);
+        
+        /* Sync the directory to ensure rename is persisted */
+        if (fsyncFileDir(server.rdb_filename) == -1) {
+            serverLog(LL_WARNING,
+                "Failed trying to sync DB directory %s in "
+                "MASTER <-> REPLICA synchronization: %s",
+                server.rdb_filename, strerror(errno));
+            cancelReplicationHandshake(1);
+            return;
```

Comment:
I'm not sure if the error is recoverable, if not the directory will not be persisted. As it's very unlikely, I think it's also ok to just log a warning and fix if there're relative issues fed back.

Reply:
@oranagra I assume the supporting argument for your suggestion is that in this case is that we don't store the file for persistence purposes, but more as a transient step before loading it. Is that what you meant?

Reply:
It seems RDB is also stored in `server.rdb_filename`, the same as replication fullsync

Reply:
@uvletter I'm not sure I understand what you mean, but I do agree with @oranagra that this should not fail replication.

Reply:
I initially thought replication and RDB persistence both store in `server.rdb_filename`, a failed replication `fsync` may affect RDB persistence. After a rethink I agree with your opinion, replication and RDB persistence are two individual functions, and only RDB persistence should concern about the `fsync` failures.

Reply:
One more interesting thing I found is that the `fsync` at https://github.com/redis/redis/blob/ed29d634b3844d2c6af099e22f3bd6bf316edab9/src/replication.c#L2126 may also be a "best effort" attempt, which means `cancelReplicationHandshake(1)` is unnecessary. https://github.com/redis/redis/commit/784b93087c622999e225735c0c76d50241941022 mentions `rdb_fsync_range` is just to avoid the blocking at the final `rename`. Should I correct it together？

Reply:
i think i'd like to retract my comment.
when using disk-based replication, the assumption is that the RDB file we received is also meant to serve for persistence.
arguably, only when `server.saveparamslen>0`.

on the other hand, after reading about all the other databases that completely ignore directory fsync failure, i'm not sure we wanna take any other approach.